### ISSUE TITLE
Why does a local install attempt of ruby require write access to a system directory?

### ISSUE THREAD

**Issue by kenbod** [rbenv/ruby-build#755]
Hi,

I tried to install ruby 2.2.2 using the command "rbenv install 2.2.2" today on an Ubuntu 12.04 machine using ruby-build 20150502 and it fails near the end of the install with this error message:

```
installing bundle gems:       /home/kena/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0 (build_info, cache, doc, extensions, gems, specifications)
/tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/installer.rb:645:in `verify_gem_home': You don't have write permissions for the /usr/local/rvm/gems/ruby-2.0.0-p247 directory. (Gem::FilePermissionError)
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/installer.rb:765:in `pre_install_checks'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/installer.rb:231:in `install'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/resolver/specification.rb:96:in `install'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/request_set.rb:257:in `block in install_into'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/request_set.rb:249:in `each'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/request_set.rb:249:in `install_into'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/request_set.rb:141:in `install'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems/dependency_installer.rb:394:in `install'
	from /tmp/ruby-build.20150503094151.31762/ruby-2.2.2/lib/rubygems.rb:558:in `install'
	from ./tool/rbinstall.rb:722:in `block (2 levels) in <main>'
	from ./tool/rbinstall.rb:721:in `each'
	from ./tool/rbinstall.rb:721:in `block in <main>'
	from ./tool/rbinstall.rb:757:in `call'
	from ./tool/rbinstall.rb:757:in `block in <main>'
	from ./tool/rbinstall.rb:754:in `each'
	from ./tool/rbinstall.rb:754:in `<main>'
```

The "/home/kena/.rbenv/versions/2.2.2" directory is where I'm trying to install for this particular build but as you can see the build is failing because I don't have write permission for the "/usr/local/rvm/gems/ruby-2.0.0-p247/" directory. That directory, however, is part of the system-installed ruby and I don't want my 2.2.2 install to have anything to do with that version of ruby, I just want to install ruby 2.2.2 as a stand-alone install in my user directory.

Why does the installation process require write access to a directory that is part of the system-installed version of ruby?

**Comment by seitensei** [comment]
Is 2.0.0-p247 installed using rvm? Try removing any references to rvm within your environment, if possible.

I don't use it myself, but it seems that rbenv doesn't like rvm, according to the readme.

**Comment by kenbod** [comment]
Yeah, I'm unable to remove the rvm-installed version on this particular machine, since it is a machine used by multiple users and most of them just use that version of ruby to get their work done. I'm trying to upgrade to 2.2.2 and switch over to using rbenv instead of rvm. The only thing preventing me is that the installation process somehow determines that there is an rvm-installed version of ruby on the machine and wants write access to its gems directory. My question is why does the installer want to write into that directory in the first place? I'm just trying to install a stand-alone copy of ruby within rbenv's directories. If I could just get it to ignore the rvm-installed version, I feel like the install process would otherwise finish successfully. Thanks.

**Comment by mislav** [comment]
Does it help if you reset RUBYOPT, GEM_HOME, GEM_PATH environment variables to blank values before attempting ruby-build install?

**Comment by mislav** [comment]
Hi @kenbod I strongly suspect that your issue is due to RVM present on the machine and some environment misconfiguration around RubyGems, per my above comment. Do you think this is a bug of ruby-build project or can I close this?

**Comment by kenbod** [comment]
Hi Mislav,

I have been away at two conferences and wasn't able to work on this 
while I was away. I arrive home late on Sunday night and will try out 
your suggestion on Monday and provide you with feedback on whether it 
had an impact. I would appreciate keeping this open just a little bit 
longer.

Apologies for the delay,

Ken

Mislav Marohnić wrote:

**Comment by mislav** [comment]
Alright no rush!

**Comment by kenbod** [comment]
Darn. I nuked all rvm-related and gem related environment variables and it looked like I was going to make it this time. And, then, it failed with with a call to chmod:

installing default gems:      /home/kena/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0 (build_info, cache, doc, extensions, gems, specifications)
                              bigdecimal 1.2.6
                              io-console 0.4.3
                              json 1.8.1
                              psych 2.0.8
                              rake 10.4.2
                              rdoc 4.2.0
installing bundle gems:       /home/kena/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0 (build_info, cache, doc, extensions, gems, specifications)
                              power_assert-0.2.2.gem
                              minitest-5.4.3.gem
                              test-unit-3.0.8.gem
./tool/rbinstall.rb:728:in `chmod': Operation not permitted @ chmod_internal - / (Errno::EPERM)
	from ./tool/rbinstall.rb:728:in `block in <main>'
	from ./tool/rbinstall.rb:757:in `call'
	from ./tool/rbinstall.rb:757:in `block in <main>'
	from ./tool/rbinstall.rb:754:in `each'
	from ./tool/rbinstall.rb:754:in `<main>'
make: *** [do-install-all] Error 1

I'm invoking this as "rbenv install 2.2.2". I want to avoid invoking it as "sudo rbenv install 2.2.2" since that would defeat the purpose of a local ruby build. Any suggestions?

**Comment by mislav** [comment]
Please try

    find `rbenv root`/versions ! -perm -u+w

to find all files and directories that you might not have write access to

**Comment by kenbod** [comment]
Okay, I ran that command, but, unfortunately, it failed again in the same spot. Is there anything I can do to provide more information? If not, I'll give up and just use rvm on that particular system.

**Comment by mislav** [comment]
> Okay, I ran that command

What was its output? The command is just supposed to print offending files. It doesn't fix anything.

> it failed again in the same spot

What failed, `rbenv install`? Naturally, if you haven't fixed permissions in the meantime. Can we first focus on finding which directory you don't have permission to write to, then `chmod`ing it, then trying install again?

**Comment by kenbod** [comment]
The versions directory is empty. I haven't been able to complete a successful install. So, instead, I ran the command above on the tmp directory that is created when I run the install command. And, yes, this chmod error is occurring when I run the command "rbenv install 2.2.2".

The tmp directory that is created when I run that command is located here (with a new directory each time I run the command, of course):

/tmp/ruby-build.20150528130322.18488/

When I run your find command on that directory, I get:

ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/hoe/minitest.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/mock.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/autorun.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/hell.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/unit.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/spec.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/parallel.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/pride_plugin.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/assertions.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/test.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/benchmark.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/expectations.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest/pride.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/lib/minitest.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/.autotest
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/Rakefile
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/README.txt
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/History.txt
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/test_minitest_spec.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/test_minitest_reporter.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/test_minitest_mock.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/test_minitest_benchmark.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/test_minitest_unit.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/test/minitest/metametameta.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/design_rationale.rb
ruby-build.20150528130322.18488/ruby-2.2.2/gems/minitest-5.4.3/Manifest.txt

So, this means the culprit during the install is the minitest gem. The last bit of log before the chmod error reads:

2642 installing bundle gems:       /home/kena/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0 (build_info,      cache, doc, extensions, gems, specifications)
2643                               power_assert-0.2.2.gem
2644                               minitest-5.4.3.gem
2645                               test-unit-3.0.8.gem
2646 ./tool/rbinstall.rb:728:in `chmod': Operation not permitted @ chmod_internal - / (Errno::EPERM)
2647         from ./tool/rbinstall.rb:728:in `block in <main>'
2648         from ./tool/rbinstall.rb:757:in `call'
2649         from ./tool/rbinstall.rb:757:in `block in <main>'
2650         from ./tool/rbinstall.rb:754:in `each'
2651         from ./tool/rbinstall.rb:754:in `<main>'
2652 make: *** [do-install-all] Error 1

If I look at the gems directory where these gems are installed, I see:

epic-analytics:gems $ ls -l
total 220
-rw-r--r-- 1 kena kena     50 Dec 12 06:00 bundled_gems
drwxrwxrwx 4 kena kena   4096 May 28 13:16 minitest-5.4.3/
-rw-r--r-- 1 kena kena  68608 Nov 11  2014 minitest-5.4.3.gem
drwxrwxrwx 4 kena kena   4096 May 28 13:16 power_assert-0.2.2/
-rw-r--r-- 1 kena kena  12800 Nov 22  2014 power_assert-0.2.2.gem
drwxrwxrwx 6 kena kena   4096 May 28 13:16 test-unit-3.0.8/
-rw-r--r-- 1 kena kena 119808 Dec 12 05:54 test-unit-3.0.8.gem

and inside the minitest-5.4.3 directory, I see:

-r--r--r-- 1 kena kena  1918 May 28 13:16 design_rationale.rb
-r--r--r-- 1 kena kena 32709 May 28 13:16 History.txt
drwxrwxrwx 4 kena kena  4096 May 28 13:16 lib/
-r--r--r-- 1 kena kena   617 May 28 13:16 Manifest.txt
-r--r--r-- 1 kena kena  2003 May 28 13:16 Rakefile
-r--r--r-- 1 kena kena 19847 May 28 13:16 README.txt
drwxrwxrwx 3 kena kena  4096 May 28 13:16 test/

So, indeed, I do not have write permission on these files.

Hope this helps! And thanks for your time!

**Comment by mislav** [comment]
Thanks for the output. I've looked into this and it turns out that minitest files are always read-only for some reason after a Ruby 2.1.2+ installation. I've searched for this on the web and I've found rumors that this is a RubyGems bug, but nothing substantial.

Luckily, this is easily fixable. Your problem ocurred because you already had a Ruby 2.2.2 version installed in rbenv and you were overwriting it. However, the read-only files couldn't be overwritten, and the installation failed. The solution should be to either fix the permissions and try again, but if you don't want to keep the old Ruby 2.2.2 files, then just delete the version altogether and try the install again:

```
rm -rf `rbenv root`/versions/2.2.2
rbenv install 2.2.2
```

**Comment by kenbod** [comment]
I wish that was the case. I so want this to work.

But, as I mentioned, this is the first time I'm running the "rbenv install 2.2.2" command.

My versions directory is empty:

epic-analytics:versions $ cd `rbenv root`/versions/
epic-analytics:versions $ ls -la
total 8
drwxrwxr-x  2 kena kena 4096 May 28 13:16 ./
drwxr-xr-x 12 kena kena 4096 May  3 09:27 ../

**Comment by mislav** [comment]
Thanks for checking. Then I was on the wrong track. It looks like the minitest files weren't the culprit at all— they *are* read-only, but they aren't causing your failure.

I'll have to boot up Ubuntu to try to reproduce when I have more time. There is [a similar SO question](http://stackoverflow.com/questions/27832732/error-running-rvm-make-install#comment44071533_27832732) where the answer was to uninstall a misplaced Ruby installation. Both the SO poster and you have the same setup for the problem: you were both using RVM prior to this.

Is `/usr/local/rvm` completely gone from your system? I have no idea what actually causes the error, mind you; I just though I would ask.

**Comment by kenbod** [comment]
Thanks for this! No, /usr/local/rvm is still there. This is on a machine that has 20 odd users across two different research groups. Those other users all make use of the system installed rvm, so I can't touch that directory (unfortunately).

**Comment by mislav** [comment]
Try changing the temporary compile directory to somewhere under your $HOME, e.g.:

```
mkdir ~/ruby-tmp
TMPDIR=~/ruby-tmp rbenv install 2.2.2
```

You can delete that directory afterwards.

Also, tell me the output of following:

```
type rvm
env | grep 'RUBY\|GEM'
```

**Comment by kenbod** [comment]
The output of the env command is empty since I'm running this command before I do the install to remove all references to 'rvm' in my environment.

export rvm_env_string="" ; export rvm_ruby_string="" ; export GEM_HOME=""; export rvm_bin_path=""; export rvm_path=""; export rvm_prefix=""; export MY_RUBY_HOME=""; export rvm_version=""; export GEM_PATH=""; export RUBY_VERSION=""

So, after I run that command, the output of the env command is:

env | grep 'RUBY\|GEM'
GEM_HOME=
MY_RUBY_HOME=
GEM_PATH=
RUBY_VERSION=

The output of the 'type rvm' command is

rvm is a function
rvm ()
{
    typeset result current_result;
    export -a rvm_ruby_args > /dev/null 2> /dev/null;
    rvm_ruby_args=();
    __rvm_teardown_if_broken;
    __rvm_cli_posix_check || return $?;
    __rvm_cli_load_rvmrc || return $?;
    __rvm_cli_version_check "$@" || return $?;
    __rvm_cli_autoupdate "$@" || return $?;
    __rvm_path_match_gem_home_check;
    __rvm_initialize;
    __rvm_setup;
... it goes on for quite some time ....
    __rvm_cli_autoreload;
    if (( __local_rvm_trace_flag > 0 )); then
        set +o verbose;
        set +o xtrace;
        [[ -n "${ZSH_VERSION:-""}" ]] || set +o errtrace;
    fi;
    return ${result:-0}
}

I will try running the install command with the different TMPDIR now. Stay tuned.

**Comment by kenbod** [comment]
Okay, the install with a new tmp directory just finished. It fails in the same place:

installing bundle gems:       /home/kena/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0 (build_info, cache, doc, extensions, gems, specifications)
                              power_assert-0.2.2.gem
                              minitest-5.4.3.gem
                              test-unit-3.0.8.gem
./tool/rbinstall.rb:728:in `chmod': Operation not permitted @ chmod_internal - / (Errno::EPERM)
	from ./tool/rbinstall.rb:728:in `block in <main>'
	from ./tool/rbinstall.rb:757:in `call'
	from ./tool/rbinstall.rb:757:in `block in <main>'
	from ./tool/rbinstall.rb:754:in `each'
	from ./tool/rbinstall.rb:754:in `<main>'
make: *** [do-install-all] Error 1

FYI.

**Comment by mislav** [comment]
If `type rvm` isn't empty, that means RVM is still effective in the current environment, even though you've nulified the environment variables. It *might* be messing with our install somehow, although I'm not sure how, or why is installation trying to chmod the root `/` directory. Maybe it has something to do with you nulifying the environment. Maybe you should bring the environment values back, and see how it affects the error message.

The only way to not unload RVM would be to prevent it from loading in the current shell in the first place. Unfortunately, RVM loading might be enabled on the system level, in files like `/etc/bash.bashrc` or `/etc/profile`. You might not have privileges to change those files.

If you can't remove RVM from this shared environment, maybe you should simply continue using RVM instead of rbenv? It might be easier for you.

* * *

BTW: when you paste code or terminal output, like you did many times above, it helps if you enclose the pasted content in three backticks on each side:

    ```
    {pasted content}
    ```

GitHub then ensures that the plain text formatting of the output stays preserved.

**Comment by secondstreetmedia** [comment]
I know this is a couple of months old, but if this helps you at all, I was having the same problem/error when attempting to use ruby-install (https://github.com/postmodern/ruby-install) with RVM still installed.  Since I was planning to switch from RVM to Chruby anyway, I uninstalled RVM following these steps (from https://coderwall.com/p/8r4uog/install-chruby-to-replace-rvm-on-os-x with one additional file name in Step 3):

Step 1.
```
~>>rvm implode
```

Step 2.
```
~>>gem uninstall rvm
```

Step 3.
Remove references to rvm in any .bashrc, .bash_profile, .profile, .zprofile, .zshenv or .zsrsh files (there were many references).

After Step 1 or Step 2 (I don't recall which) there were messages regarding certain files and the .rvm directory that I would have to delete manually, which I did.

I exited the terminal, restarted and reinstalled ruby 2.2.2 and was able to do so successfully.  So, it appears to be related to RVM, although having done this ruby install before installing Chruby, I can't be certain it wouldn't have happened again had Chruby already been installed.  Perhaps there is a way to redo the steps regarding RVM that I have undone to see where the issue begins.

Hope it helps,
Tim

**Comment by mislav** [comment]
Thanks @secondstreetmedia! :bow:

**Comment by mislav** [comment]
Closing since this was caused by RVM and there's not much we can do from ruby-build.

**Comment by vaibhav13** [comment]
I know this issue is closed. Just wanted to add something related to this. I was also facing this issue while installing ruby 2.2.6. I already had 2.2.2 installed locally. While installing 2.2.6 I was getting this error: 
./tool/rbinstall.rb:728:in `chmod': Operation not permitted @ chmod_internal - /usr/local/rvm/gems/ruby-2.2.2/extensions/x86_64-darwin-15/ (Errno::EPERM)

I resolved this issue by removing 2.2.2 version first and then installing 2.2.6 and then 2.2.2.
sudo rvm remove 2.2.2
rvm install 2.2.6
rvm install 2.2.2
